src/unix/pty.cc

```diff
@@ -99,27 +102,106 @@ int pthread_fchdir_np(int fd) API_AVAILABLE(macosx(10.12));
+      0,                            // Unlimited queue
+      1,                            // Only one thread will use this initially
+      [th](Napi::Env) {   // Finalizer used to clean threads up
+        th->join();
+        delete th;
+      });
+  *th = std::thread([tsfn = std::move(tsfn), pid] {
+    auto callback = [](Napi::Env env, Napi::Function cb, ExitEvent *exit_event) {
+      cb.Call({Napi::Number::New(env, exit_event->exit_code),
+               Napi::Number::New(env, exit_event->signal_code)});
+      delete exit_event;
+    };
 
-struct pty_baton {
-  Nan::Persistent<v8::Function> cb;
-  int exit_code;
-  int signal_code;
-  pid_t pid;
-  uv_async_t async;
-  uv_thread_t tid;
-};
+    int ret;
+    int stat_loc;
+#if defined(__APPLE__)
+    // Based on
+    // https://source.chromium.org/chromium/chromium/src/+/main:base/process/kill_mac.cc;l=35-69?
+    int kq = HANDLE_EINTR(kqueue());
+    struct kevent change = {0};
+    EV_SET(&change, pid, EVFILT_PROC, EV_ADD, NOTE_EXIT, 0, NULL);
+    ret = HANDLE_EINTR(kevent(kq, &change, 1, NULL, 0, NULL));
+    if (ret == -1) {
+      if (errno == ESRCH) {
+        // At this point, one of the following has occurred:
+        // 1. The process has died but has not yet been reaped.
+        // 2. The process has died and has already been reaped.
+        // 3. The process is in the process of dying. It's no longer
+        //    kqueueable, but it may not be waitable yet either. Mark calls
+        //    this case the "zombie death race".
+        ret = HANDLE_EINTR(waitpid(pid, &stat_loc, WNOHANG));
+        if (ret == 0) {
+          ret = kill(pid, SIGKILL);
+          if (ret != -1) {
+            HANDLE_EINTR(waitpid(pid, &stat_loc, 0));
+          }
+        }
+      }
+    } else {
+      struct kevent event = {0};
+      ret = HANDLE_EINTR(kevent(kq, NULL, 0, &event, 1, NULL));
+      if (ret == 1) {
+        if ((event.fflags & NOTE_EXIT) &&
+            (event.ident == static_cast<uintptr_t>(pid))) {
+          // The process is dead or dying. This won't block for long, if at
+          // all.
+          HANDLE_EINTR(waitpid(pid, &stat_loc, 0));
+        }
+      }
+    }
+#else
+    while (true) {
+      errno = 0;
+      if ((ret = waitpid(pid, &stat_loc, 0)) != pid) {
+        if (ret == -1 && errno == EINTR) {
+          continue;
+        }
+        if (ret == -1 && errno == ECHILD) {
+          // XXX node v0.8.x seems to have this problem.
+          // waitpid is already handled elsewhere.
+          ;
+        } else {
+          assert(false);
+        }
+      }
+      break;
+    }
+#endif
+    ExitEvent *exit_event = new ExitEvent;
+    if (WIFEXITED(stat_loc)) {
+      exit_event->exit_code = WEXITSTATUS(stat_loc); // errno?
+    }
+    if (WIFSIGNALED(stat_loc)) {
+      exit_event->signal_code = WTERMSIG(stat_loc);
+    }
+    auto status = tsfn.BlockingCall(exit_event, callback); // In main thread
+    assert(status == napi_ok);
+
+    tsfn.Release();
```

Comment:
this fails for me (electron 28, node 20, gcc/libstdc++ 13)
```
make: Entering directory '/home/lauren/aports/testing/code-oss/src/vscode-1.85.0/node_modules/node-pty/build'
  CXX(target) Release/obj.target/pty/src/unix/pty.o
../src/unix/pty.cc: In lambda function:
../src/unix/pty.cc:193:17: error: passing 'const Napi::ThreadSafeFunction' as 'this' argument discards qualifiers [-fpermissive]
  193 |     tsfn.Release();
      |     ~~~~~~~~~~~~^~
In file included from ../../node-addon-api/napi.h:2725,
                 from ../src/unix/pty.cc:21:
../../node-addon-api/napi-inl.h:5173:20: note:   in call to 'napi_status Napi::ThreadSafeFunction::Release()'
 5173 | inline napi_status ThreadSafeFunction::Release() {
      |                    ^~~~~~~~~~~~~~~~~~
make: *** [pty.target.mk:127: Release/obj.target/pty/src/unix/pty.o] Error 1
```

Reply:
You're using an ancient version of `node-addon-api` package. In [node-addon-api@7.0.0 's napi-inl.h](https://cdn.jsdelivr.net/npm/node-addon-api@7.0.0/napi-inl.h) :

```cpp
inline napi_status ThreadSafeFunction::Release() const {
```

But this method in [node-addon-api@4.3.0](https://cdn.jsdelivr.net/npm/node-addon-api@4.3.0/napi-inl.h) doesn't have `const`.

Please check and ensure you're using the correct version (there's `"node-addon-api": "^7.0.0"` in package.json).

Reply:
thanks a lot